pkg/ice/agent.go

```diff
@@ -644,11 +645,55 @@ func (a *Agent) handleInboundControlling(m *stun.Message, localCandidate, remote
 	}
 }
 
+// handleNewPeerReflexiveCandidate adds an unseen remote transport address
+// to the remote candidate list as a peer-reflexive candidate.
+func (a *Agent) handleNewPeerReflexiveCandidate(local *Candidate, remote net.Addr) error {
+	var ip net.IP
+	var port int
+
+	switch addr := remote.(type) {
+	case *net.UDPAddr:
+		ip = addr.IP
+		port = addr.Port
+	case *net.TCPAddr:
+		ip = addr.IP
+		port = addr.Port
+	default:
+		return errors.Errorf("unsupported address type %T", addr)
+	}
+
+	pflxCandidate, err := NewCandidatePeerReflexive(
+		local.NetworkType.String(), // assume, same as that of local
+		ip,
+		port,
+		"", // unknown at this moment. TODO: need a review
+		0,  // unknown at this moment. TODO: need a review
+	)
+
+	if err != nil {
+		return errors.Wrapf(err, "failed to create peer-reflexive candidate: %v", remote)
+	}
+
+	// Add pflxCandidate to the remote candidate list
+	// TODO: review if this wouldn't cause any race
+	networkType := pflxCandidate.NetworkType
```

Comment:
Are you able to call `AddRemoteCandidate` directly here? It handles all the locking/modifying for you

```
return a.AddRemoteCandidate(pflixCandidate)
```

Reply:
Just confirmed, the thread has already in agent.run() routine.  Calling another agent.run() would be a dead lock.  (I wonder if there's away to allow "recursivce mutex locking".)

Reply:
Oops sorry my mistake!

I would make a private version `addRemoteCandidate` that assumes you have the lock so it could look like this.

```
// AddRemoteCandidate adds a new remote candidate
func (a *Agent) AddRemoteCandidate(c *Candidate) error {
  return a.run(func (agent *Agent) {
    a.addRemoteCandidate(c)
  })
}

// addRemoteCandidate assumes you are holding the lock (must be execute using a.run)
func (a *Agent) addRemoteCandidate(c *Candidate) error {
        networkType := c.NetworkType
        set := agent.remoteCandidates[networkType]

        for _, candidate := range set {
            if candidate.Equal(c) {
                return
            }
        }

        set = append(set, c)
        agent.remoteCandidates[networkType] = set
}
```

Reply:
Iterating over remoteCandidates has already been done earlier by findRemoteCandidate(). So, it would redundant. Having said that, I agree using the new `addRemoteCandidate` might be idiomatically better.  Let me try.